pkg/terror/terror.go

```diff
@@ -0,0 +1,302 @@
+	switch verb {
+	case 'v':
+		if s.Flag('+') {
+			io.WriteString(s, e.Error())
+			if e.stack != nil {
+				e.stack.StackTrace().Format(s, verb)
+			}
+			return
+		}
+		fallthrough
+	case 's':
+		io.WriteString(s, e.Error())
+	case 'q':
+		fmt.Fprintf(s, "%q", e.Error())
+	}
+}
+
+// Cause implements causer.Cause defined in pingcap/errors
+// and returns the raw cause of an *Error
+func (e *Error) Cause() error {
+	return e.rawCause
+}
+
+func (e *Error) getMsg() string {
+	if len(e.args) > 0 {
+		return fmt.Sprintf(e.message, e.args...)
+	}
+	return e.message
+}
+
+// Equal checks if err equals to e.
+func (e *Error) Equal(err error) bool {
+	if error(e) == err {
+		return true
+	}
+	inErr, ok := err.(*Error)
+	return ok && e.code == inErr.code
+}
+
+// New generates a new *Error with the same class and code, and replace message with new message
+func (e *Error) New(message string) error {
+	return e.levelGeneratef(1, message)
+}
+
+// Generate generates a new *Error with the same class and code, and new arguments.
+func (e *Error) Generate(args ...interface{}) error {
+	return e.levelGeneratef(1, e.message, args...)
+}
+
+// Generatef generates a new *Error with the same class and code, and a new formatted message.
+func (e *Error) Generatef(format string, args ...interface{}) error {
+	return e.levelGeneratef(1, format, args...)
+}
+
+// levelGeneratef is an inner interface to generate new *Error
+func (e *Error) levelGeneratef(stackSkipLevel int, format string, args ...interface{}) error {
+	err := &Error{
+		code:    e.code,
+		class:   e.class,
+		scope:   e.scope,
+		level:   e.level,
+		message: format,
+		args:    args,
+		stack:   errors.NewStack(stackSkipLevel),
+	}
+	return err
+}
+
+// Delegate creates a new *Error with the same fields of the give *Error,
+// except for new arguments, it also sets the err as raw cause of *Error
+func (e *Error) Delegate(err error, args ...interface{}) error {
+	if err == nil {
+		return nil
+	}
+	err2 := &Error{
+		code:     e.code,
+		class:    e.class,
+		scope:    e.scope,
+		level:    e.level,
+		message:  fmt.Sprintf("%s: %s", e.message, err),
+		args:     args,
+		rawCause: err,
+		stack:    errors.NewStack(0),
+	}
+	return err2
+}
+
+// Annotate tries to convert err to *Error and adds a message to it.
+// This API is designed to reset Error message but keeps its original trace stack
+func Annotate(err error, message string) error {
+	if err == nil {
+		return nil
+	}
+	e, ok := err.(*Error)
+	if !ok {
+		return errors.Annotate(err, message)
+	}
+	e.message = fmt.Sprintf("%s: %s", message, e.getMsg())
+	e.args = nil
+	return e
+}
+
+// Annotatef tries to convert err to *Error and adds a message to it.
+func Annotatef(err error, format string, args ...interface{}) error {
+	if err == nil {
+		return nil
+	}
+	e, ok := err.(*Error)
+	if !ok {
+		return errors.Annotatef(err, format, args...)
+	}
+	e.message = fmt.Sprintf("%s: %s", format, e.getMsg())
```

Comment:
:thinking:

Reply:
should it  be `e.message = e.message + ": " +  fmt.Sprintf(fmt, args)`?

Reply:
nope, this keeps the same as `errors.Annotatef` we used in pingcap/errors
such as https://github.com/pingcap/errors/blob/master/format_test.go#L140-L154